### PR TITLE
Optimize password authentication.

### PR DESCRIPTION
**Pull request by riverszhang89**
Authenticate the user/password for a new connection instead of every statement. If a user's password has changed, force re-authentication on all existing connections.

In addition, make the number of PBKDF2 iterations configurable.

### CODE REVIEW

File: bdb/llmeta.c [hunk 1]
```diff
@@ -8978,10 +8998,11 @@ int bdb_user_password_set(tran_type *tran, char *user, char *passwd)
         return rc;
     key.passwd.file_type = htonl(LLMETA_USER_PASSWORD_HASH);
     passwd_hash data;
-    data.ver = 1;
+    data.ver = 2;
+    data.niterations = gbl_pbkdf2_iterations;
```
[the remark this answers is not in the document]
Comment (riverszhang89) [reply bdb/llmeta.c]: Gotcha. Might need to htonl data.ver as well?

[the remark this answers is not in the document]
Comment (akshatsikarwar) [reply bdb/llmeta.c]: Not really. Its 1 byte.

File: bdb/llmeta.c [hunk 2]
```diff
@@ -8920,6 +8921,21 @@ static int llmeta_get_user_passwd(char *user, llmetakey_t type, void ***out)
 }
 #define ITERATIONS_V0 1000
 #define ITERATIONS_V1 16 * 1024
+#define ITERATIONS_MIN ITERATIONS_V0
```
[the remark this answers is not in the document]
Comment (riverszhang89) [reply bdb/llmeta.c]: The default is 16K. And it turns out to be a bit expensive. We're gonna need a smaller number in production.

File: db/sqlinterfaces.c [hunk 3]
```diff
@@ -3668,8 +3668,14 @@ static int check_sql_access(struct sqlthdstate *thd, struct sqlclntstate *clnt)
     if (gbl_check_access_controls) {
         check_access_controls(thedb);
         gbl_check_access_controls = 0;
+        /* Force all clients to  reauthenticate if passwords have changed. */
+        clnt->authenticated = 0;
```
Comment (nirbhayc) [inline db/sqlinterfaces.c]: How will this force all clients to re-authenticate?

Comment (riverszhang89) [reply db/sqlinterfaces.c]: There's an if branch right below this line, checking if clnt->authenticated is true. If not, perform password check.

Comment (nirbhayc) [reply db/sqlinterfaces.c]: Right, but that only re-auth's this client.

Comment (riverszhang89) [reply db/sqlinterfaces.c]: Once gbl_check_access_controls is set to true, all clients will invalidate the flag. gbl_check_access_controls  is set to true when 1) authentication has been enabled; 2) a user's password has changed.

Comment (riverszhang89) [reply db/sqlinterfaces.c]: Yes you're right - only the 1st client who gets in here will re-authenticate.

**Pull request merged**